### PR TITLE
(To Be discussed) (RNN) feature: burn in steps

### PR DESCRIPTION
**Pull request by YuriCat**
Burn-in is effective (or necessary in some cases) to train RNNs.

### CODE REVIEW

File: handyrl/train.py [hunk 1]
```diff
@@ -163,7 +164,14 @@ def forward_prediction(model, hidden, batch, args):
...
+                model.eval()
+                with torch.no_grad():
+                    outputs_= model(obs, hidden_)
+            else:
+                if not model.training:
+                    model.train()
+                outputs_ = model(obs, hidden_)
             for k, o in outputs_.items():
                 if k == 'hidden':
                     next_hidden = outputs_['hidden']
```
Comment (ikki407) [inline handyrl/train.py:177]: BTW, this can be `next_hidden = o` ? I don't care that though...

File: handyrl/train.py [hunk 2]
```diff
@@ -288,16 +296,18 @@ def select_episode(self):
...
                 break
         ep = self.episodes[ep_idx]
-        turn_candidates = 1 + max(0, ep['steps'] - self.args['forward_steps'])  # change start turn by sequence length
-        st = random.randrange(turn_candidates)
-        ed = min(st + self.args['forward_steps'], ep['steps'])
+        trained_steps = self.args['forward_steps'] - self.args['burn_in_steps']
+        turn_candidates = 1 + max(0, ep['steps'] - trained_steps)  # change start turn by sequence length
+        st_train = random.randrange(turn_candidates)
+        ed = min(st_train + trained_steps, ep['steps'])
+        st = max(0, st_train - self.args['burn_in_steps'])
```
Comment (ikki407) [inline handyrl/train.py:303]: I wonder which of the following two is better.

When forward_steps=24 and burn_in_steps=8, 
1. use first 8 samples of 24 samples in batch for burn-in (i.e., training with 16 samples)
2. use 8 samples for burn-in and 24 samples for training respectively  (i.e. batch size is 32)
    3. can this separate the GPU calculation between burn-in and training in `forward_prediction`?

At first, I have been thought it is the second behavior.

Comment (YuriCat) [reply handyrl/train.py:303]: Yes, feautre/burn_in_step_naive_forward_steps is the correct branch!

https://github.com/YuriCat/HandyRL/compare/feature/burn_in_steps...feautre/burn_in_step_naive_forward_steps

**Comment by ikki407** [comment]
I reviewed new version of naive forward step. LGTM.

File: handyrl/train.py (hunk 1 above)
Comment (YuriCat) [reply handyrl/train.py:166]: I belive so. I will create chore PR.

**Comment by ikki407** [comment]
👍

**Comment by ikki407** [comment]
I like this PR so I will merge it. I hope that more information about RNNs will be added in HandyRL.

**Pull request merged**